File: 240425/네 방향 탈출 가능 여부 판별하기/determine-escapableness-with-4-ways.cpp
#include <iostream>
#include <queue>

#define MAX_N 100
#define MAX_M 100

using namespace std;

int n, m;
int grid[MAX_N][MAX_M];

queue<pair<int, int>> q;
bool visited[MAX_N][MAX_M];

bool in_range(int r, int c) {
    return 0 <= r && r < n && 0 <= c && c < m;
}

bool can_go(int r, int c) {
    return in_range(r, c) && grid[r][c] && !visited[r][c];
}

void bfs() {

    int dr[4] = {1, 0, -1, 0};
    int dc[4] = {0, 1, 0, -1};

    while (!q.empty()) {
        pair<int, int> pos = q.front();
        int r = pos.first, c = pos.second;
        q.pop();

        for (int i = 0; i < 4; i++) {
            int nr = r + dr[i], nc = c + dc[i];
            if (can_go(nr, nc)) {
                q.push(make_pair(nr, nc));
                visited[nr][nc] = true;
            }
        }
    }
}

int main() {
    cin >> n >> m;
    for (int r = 0; r < n; r++) {
        for (int c = 0; c < m; c++) {
            cin >> grid[r][c];
        }
    }

    q.push(make_pair(0, 0));
    visited[0][0] = true;

    bfs();
    cout << visited[n - 1][m - 1];

    return 0;
}

File: 240425/네 방향 탈출 가능 여부 판별하기/determine-escapableness-with-4-ways.py
from queue import Queue

n, m = map(int, input().split())
grid = [list(map(int, input().split())) for _ in range(n)]
visited = [[False] * m for _ in range(n)]

q = Queue()

drs, dcs = [1, 0, -1, 0], [0, 1, 0, -1]

def in_range(r, c):
    return 0 <= r < n and 0 <= c < m

def bfs():
    q.put((0, 0))
    visited[0][0] = True

    while not q.empty():
        r, c = q.get()
        for dr, dc in zip(drs, dcs):
            nr, nc = r + dr, c + dc
            if in_range(nr, nc) and grid[nr][nc] == 1 and not visited[nr][nc]:
                visited[nr][nc] = True
                q.put((nr, nc))

bfs()
print(1 if visited[n - 1][m - 1] else 0)